File: python/Edit Distance.py
class Solution:
    # @param {string} word1
    # @param {string} word2
    # @return {integer}
    def minDistance(self, word1, word2):
        self.d = [[-1 for j in range(len(word2)+1)] for i in range(len(word1)+1)]
        # d[i][j] for the distance of word1[:i], word2[:j]
        return self.dp(len(word1), len(word2), word1, word2)
    def dp(self, i, j, word1, word2):
        if self.d[i][j] != -1:
            return self.d[i][j]
        if i == 0:
            self.d[i][j] = j
        elif j == 0:
            self.d[i][j] = i
        elif word1[i-1] == word2[j-1]:
            self.d[i][j] = self.dp(i-1,j-1,word1,word2)
        else:
            self.d[i][j] = min(self.dp(i-1,j-1,word1,word2), self.dp(i-1,j,word1,word2), self.dp(i,j-1,word1,word2))+1
            # replace, delete, insert
        return self.d[i][j]

word1 = 'ab'
word2 = 'bcasda'
print(Solution().minDistance(word1, word2))
        

File: cpp/Edit Distance.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
	int minDistance(const string &word1, const string &word2) {
		if (word1.length() < word2.length())
			return minDistance(word2, word1);
		
		vector<int> f(word2.length() + 1); // shorter
		int upper_left = 0;

		for (size_t i = 0; i <= word2.size(); ++i)
			f[i] = i;

		for (size_t i = 1; i <= word1.size(); ++i) {
			upper_left = f[0];
			f[0] = i;

			for (size_t j = 1; j <= word2.size(); ++j) {
				int upper = f[j];
				if (word1[i - 1] == word2[j - 1])
					f[j] = upper_left;
				else 
					f[j] = 1 + min(upper_left, min(f[j], f[j-1]));
				upper_left = upper;
			}
		}

		return f[word2.length()];
	}
};

/*class Solution {
public:
	int minDistance(const string &word1, const string &word2) {
		const size_t n = word1.size();
		const size_t m = word2.size();
		vector<vector<int> > f(n + 1, vector<int>(m + 1, 0));
		for (size_t i = 0; i <= n; ++i)
			f[i][0] = i;
		for (size_t j = 0; j <= m; ++j)
			f[0][j] = j;

		for (size_t i = 1; i <= n; ++i) {
			for (size_t j = 1; j <= m; ++j) {
				if (word1[i - 1] == word2[j - 1]) {
					f[i][j] = f[i - 1][j - 1];
				}
				else {
					// f[i - 1][j]: delete word1[-1]
					// f[i][j - 1]: delete word2[-1]
					// f[i - 1][j - 1]: change word1[-1] to word2[-1]
					int mn = min(f[i - 1][j], f[i][j - 1]);
					f[i][j] = 1 + min(f[i - 1][j - 1], mn);
				}
					
			}
		}
	}
};*/